### ISSUE TITLE
Stack smashing detected

### ISSUE THREAD

**Issue by kremso** [ohler55/oj#545]
Hello,

here's a "minimal" ruby script which causes a "stack smashing detected" error. 

```
$ ruby test.rb
*** stack smashing detected ***: <unknown> terminated
fish: Job 3, “ruby test.rb” terminated by signal SIGABRT (Abort)
```

The data inside the script is intentionally nonsensical. It was originally an elasticsearch query, but I've mangled the key names and values to not leak private data.
The original query (ergo json data) was much longer, but I was able to remove large parts of it, while still being able to reproduce the error. This is the shortest version I could find -- shortening the data (either key name, or value) just by 1 character, causes the error to disappear.

Tested on ruby 2.6.3 and latest released oj 3.7.12 as well as with the current master branch, tested on linux and macos, and the error reproduces consistently.

```
require 'json'
require 'oj'

Oj.mimic_JSON

obj = {
  ab: {
    cbbb: {
      tilbeb: [
        {
          coob: {
            uijwts: [
              {
                prrrrr: {
                  yakj: "pvebbx",
                  lbhqy: {
                    uhyw: {
                      uijwts: [
                        {
                          jangi: {
                            ubentg7haineued8atnr8w: {
                              abc: "uejdncbncnamnasdasdasdasd",
                              cde: "skfjskdfjskdfjsdkfjsdkfjs"
                            }
                          }
                        }
                      ]
                    }
                  }
                }
              },
              {
                kdncg: {
                  lvbnt8b9ounv: {
                    qk: 9
                  }
                }
              }
            ],
            jenfjbhe: {}
          }
        }
      ]
    }
  },
  ijbh: {
    jsnbrpbnunt: {
      b88dibalbvp: {
        mnbvd: "9uhbqlpiev"
      }
    },
    ncnwkl: {
      ksdfsf: {
        mjln: "mnklkn"
      },
      kbrh: {
        sdfn83nnalbmgnansdd: {
          uijwts: {
            ibha: {
              uijwts: [
                {
                  lnrbf: {
                    nbvtmqbhap9ebeb7btnnaw: {
                      ksb: "sdfksdfjsdfsb39242dnasddd",
                      mnm: "1293dsfnsdmfnsdfsd,fmnsd,"
                    }
                  }
                }
              ]
            }
          },
          kbrh: {
            bo8libts: {
              nag40n: {
                kyen: "sdfasnc92nsn"
              },
              kbrh: {
                nbwyu26snfcbajsdkj8: {
                  uijwts: {
                    mdfnkjsdd: {}
                  },
                  kbrh: {
                    kneahce: {
                      uijwts: {
                        kwnb: {
                          uijwts: [
                            {
                              fhfd: {
                                sfasdnfmasndfamsdnfajsmdf: false
                              }
                            }
                          ],
                          asdfsdff: [
                            {
                              cwdf: {
                                sddlkfajsdkfjabskdfjalsdkfjansdkfjf: ""
                              }
                            },
                            {
                              bsdj: {
                                sdfsjdlfkasy8kljsfsdf83jlkjfals: true
                              }
                            }
                          ]
                        }
                      },
                      kbrh: {
                        sdfsdfsddfk: {
                          uijwts: {
                            sdfsd: {
                              sdfsadf89mnlrrrqurqwvdnff: {
                                "kj": 8
                              }
                            }
                          },
                          kbrh: {
                            dkdjd: {
                              dfeteu: {
                                sdfd: "sdfasdfjlkjslrsdbb"
                              },
                              kbrh: {
                                sdfskjdfldk: {
                                  buqpen: {
                                    kjlkj: {
                                      sdflskdjfalsdkrjalwkjfsrlfjasdf: {
                                        sd: 0
                                      }
                                    }
                                  },
                                  kbrh: {
                                    sdfksljdlfksdfl: {
                                      sdfsdkfjssd: {
                                        ksdjf: "sdflsdkfjasdkaufs;ldkfjsdlf",
                                        sdfsdfsl: [5]
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

JSON.pretty_generate(obj)
```

**Comment by ohler55** [comment]
I'll look at it today or over the weekend.

**Comment by ohler55** [comment]
I just pushed a branch named `but-size` with a fix. The expected buffer size calculations for arrays was off.

**Comment by kremso** [comment]
Thanks, I can confirm that it works, both for the test script included here, and the original larger json data.

**Comment by ohler55** [comment]
Released v3.8.0.
